Include `SwapInfo` in `Payment` resulting from swap

danielgranhao opened the pull request:
Addresses #528

TODO:
- [x] Bindings

dangeross commented:
Outside of approval, I think it definitely makes more sense as part of a JOIN query and also adding an INDEX on the bolt11 column.

JssDWt commented:
I think it makes sense in the current form of the PaymentDetails. Agree with @dangeross on the JOIN, to make this a single querty. 

But tbh I'm not too fond of where the PaymentDetails are headed. It requires logic and knowledge of the different flows in order to extract the relevant information from the PaymentDetails now. It's already like that with the lnurl fields. It should probably be enum based, with a payment type and additional data for that specific type. Refactoring that may also fix https://github.com/breez/breez-sdk/issues/302. What do you think @roeierez?

ok300 commented on libs/sdk-core/src/persist/swap.rs:
```diff
@@ -245,6 +245,10 @@ impl SqliteStorage {
         self.select_single_swap("swaps.bitcoin_address = ?1", [address])
     }
 
+    pub(crate) fn get_swap_info_by_bolt11(&self, bolt11: &str) -> Result<Option<SwapInfo>> {
```
The existing `get_swap_info_by_hash` could probably be used here instead.

If we later add a `swaps_info` INDEX on `payment_hash`, it would benefit this and other callers too.

roeierez replied:
I agree we should use hash instead of bolt11. The main reason is because the bolt11 that exists on the swap is the one that the payer pays and the bolt11 that persisted with the payment is the one that the node settled. 
In case where a channel is opened as part of the swap these can be different but same hash.

danielgranhao replied:
@ok300 @roeierez Sure, makes sense. I'm working on using a single query, but I'm having issues implementing the join due to the hash being stored as a hex string on the payments table and as a blob on the swaps table. Any suggestions on how to address this?

I see the following options:
* change the payment_hash column type to string on the swaps table (requiring encoding and decoding where necessary in the current places the table it's currently used)
* persist a serialized `SwapInfo` object in `payments_external_info` in a similar fashion to how other objects are persisted at the moment (this should be fine as long as we don't care about changes to the swap information after we've redeemed the swap)

ok300 replied:
Good catch.

We could also look into the [hex](https://www.sqlite.org/lang_corefunc.html#hex) and [unhex](https://www.sqlite.org/lang_corefunc.html#unhex) core SQLITE functions. Either of them could probably already be used in the join (e.g. `WHERE hex(payment_hash_blob) == payment_hash_hex`).

If we change the DB schema to converge on a common representation of hashes (for example hex), we could also use one of these functions to migrate data "on-the-fly" in a new migration step.

What do you think?

Btw if we change the schema, it might be better to split this in 2 steps:

- first, DB migration to common hash format, which also migrates existing data
- second, add this PR on top, which can then join without having to convert blob to hex or viceversa

I can do the migration PR if you'd like.

danielgranhao replied:
> We could also look into the [hex](https://www.sqlite.org/lang_corefunc.html#hex) and [unhex](https://www.sqlite.org/lang_corefunc.html#unhex) core SQLITE functions. Either of them could probably already be used in the join (e.g. WHERE hex(payment_hash_blob) == payment_hash_hex).

Oh, I didn't know of these functions and didn't think of searching for them. Thanks!

> Alternatively, if we change the DB schema to converge on a common representation of hashes (for example hex), we could also use one of these functions to migrate data "on-the-fly" in a new migration step.

Hm, probably makes sense to switch to a common representation. Can you elaborate on why it's better to do the migration PR first? I think it would be fine to have a query using `hex()` and remove later.

I've just updated the PR to use the hex function within the payment fetching queries.

ok300 replied:
You're right, its also fine to use `hex()` for now and remove it later with the migration.

roeierez commented:
> But tbh I'm not too fond of where the PaymentDetails are headed. It requires logic and knowledge of the different flows in order to extract the relevant information from the PaymentDetails now. It's already like that with the lnurl fields. It should probably be enum based, with a payment type and additional data for that specific type. Refactoring that may also fix #302. What do you think @roeierez?

Can you elaborate? As far as I understand you are suggesting the details section to be enum but it is already an enum today (PaymentDetails). Or your mainly suggestion to add "type" to ease the differentiation?

JssDWt commented:
> Can you elaborate? As far as I understand you are suggesting the details section to be enum but it is already an enum today (PaymentDetails). Or your mainly suggestion to add "type" to ease the differentiation?

I think my suggestion would be to go for something like this:
- Maybe split a payment into a receiving side and a sending side, since they have different dynamics.
- Remove ClosedChannel from PaymentDetails, I think it doesn't belong there?
- Make the following fields part of `Payment`
```    
    pub payment_hash: String,
    pub label: String,
    pub destination_pubkey: String,
    pub payment_preimage: String,
    pub keysend: bool,
    pub bolt11: String,
```
- Have a details variant for lnurlp, lightning address, swap

Tbh I'm not sure what is the right way to make this interface for intuitive.

roeierez commented:
> > Can you elaborate? As far as I understand you are suggesting the details section to be enum but it is already an enum today (PaymentDetails). Or your mainly suggestion to add "type" to ease the differentiation?
> 
> I think my suggestion would be to go for something like this:
> 
> * Maybe split a payment into a receiving side and a sending side, since they have different dynamics.
> * Remove ClosedChannel from PaymentDetails, I think it doesn't belong there?
> * Make the following fields part of `Payment`
> 
> ```
>     pub payment_hash: String,
>     pub label: String,
>     pub destination_pubkey: String,
>     pub payment_preimage: String,
>     pub keysend: bool,
>     pub bolt11: String,
> ```
> 
> * Have a details variant for lnurlp, lightning address, swap
> 
> Tbh I'm not sure what is the right way to make this interface for intuitive.

I think the main confusion here is that the Payment is actually represents PaymentListItem.
We would like all items that affect lightning balance to belong to that list. That was the main guideline for creating one unified type.

JssDWt commented:
> I think the main confusion here is that the Payment is actually represents PaymentListItem.
> We would like all items that affect lightning balance to belong to that list. That was the main guideline for creating one unified type.

OK, it might still make sense to put the different types of payment details into their own enum variant, so you know a swap has swapdetails, an lnurl payment has lnurl payment details, etc. And you don't need to read the documentation in order to know what fields can exist when.

danielgranhao commented:
> OK, it might still make sense to put the different types of payment details into their own enum variant, so you know a swap has swapdetails, an lnurl payment has lnurl payment details, etc. And you don't need to read the documentation in order to know what fields can exist when.

I would also be in favor of something like this. Assuming the interface goes in that direction, would it make sense to still merge something like what's proposed here and work on the refactoring in another PR?

JssDWt commented:
> I would also be in favor of something like this. Assuming the interface goes in that direction, would it make sense to still merge something like what's proposed here and work on the refactoring in another PR?

Yes, makes sense

roeierez commented:
> > OK, it might still make sense to put the different types of payment details into their own enum variant, so you know a swap has swapdetails, an lnurl payment has lnurl payment details, etc. And you don't need to read the documentation in order to know what fields can exist when.
> 
> I would also be in favor of something like this. Assuming the interface goes in that direction, would it make sense to still merge something like what's proposed here and work on the refactoring in another PR?

Yes of course. Feel free to open an issue separately for discussion on changes/suggestions to the interface. We had some discussions internally and we are happy to open the discussion publicly and get more insights

ok300 reviewed (commented):
In principle looks good, but the extra complexity in how the queries are built makes them harder to understand and reason about (in future refactorings, etc).

I personally would prefer not to add the complexity in the query building, even if that means no extra join and therefore one extra query.

What does everyone else think?

roeierez commented:
> In principle looks good, but the extra complexity in how the queries are built makes them harder to understand and reason about (in future refactorings, etc).
> 
> I personally would prefer not to add the complexity in the query building, even if that means no extra join and therefore one extra query.
> 
> What does everyone else think?

I have to say I tend to agree. Sorry I  had to go all the way just to recognize it. To me it looks like the complexity doesn't worth the join benefits as the previous code was much more simpler.

danielgranhao commented:
@ok300 @roeierez Sure, makes sense. Any performance implications are also not likely to be significant. I've updated the PR and rebased to make sure the binding generation is up to date.

roeierez reviewed (approved):
LGTM! Thanks for your patient @danielgranhao

ok300 reviewed (approved):
Looking good!
